csharp/ql/lib/semmle/code/csharp/Attribute.qll

```diff
@@ -97,3 +97,45 @@ class Attribute extends TopLevelExprParent, @attribute {
 
   override string getAPrimaryQlClass() { result = "Attribute" }
 }
+
+/**
+ * An attribute with default kind, for example `[...]` on line 1 in
+ * ```csharp
+ * [MyAttribute(0)]
+ * int SomeMethod() { return 1; }
+ * ```
+ */
+class DefaultAttribute extends Attribute, @attribute_default {
+  override string getAPrimaryQlClass() { result = "DefaultAttribute" }
+}
+
+/**
+ * An attribute with return kind, for example `[...]` on line 1 in
+ * ```csharp
+ * [return: MyAttribute(0)]
+ * int SomeMethod() { return 1; }
+ * ```
+ */
+class ReturnAttribute extends Attribute, @attribute_return {
+  override string getAPrimaryQlClass() { result = "ReturnAttribute" }
```

Comment:
Lets add a top-level predicate
```ql
private string getAttributeName(Attribute a) {
  exists(string type | type = a.getType().getName() |
    if type.matches("%Attribute") then result = type.prefix(type.length() - 9) else result = type
  )
}
```

replace the body of `Attribute::toString` with

```ql
result = "[" + getAttributeName(this) + "(...)]"
```

and add an override here

```ql
override string toString() {
  result = "[return: " + getAttributeName(this) + "(...)]"
}
```

Reply:
Yes, that is a good idea!